File: algos/starter-algo-ZIPME/gamelib/algocore.py
import json

from .game_state import GameState
from .util import get_command, debug_write, BANNER_TEXT, send_command

class AlgoCore(object):
    """This class handles communication with the game itself. Your strategy should subclass it.

    Attributes:
        * config (JSON): json object containing information about the game

    """
    def __init__(self):
        self.config = None

    def on_game_start(self, config):
        """
        Override this to perform initial setup at the start of the game, based
        on the config, a json file which contains information about the game.
        """
        self.config = config

    def on_turn(self, game_state):
        """
        This step function is called every turn and is passed a string containing
        the current game state, which can be used to initialize a new GameMap
        """
        self.submit_default_turn()

    def submit_default_turn(self):
        send_command("")
        send_command("")

    # only override this function if you have a 
    def start(self):
        """ 
        Start the parsing loop.
        Python will hang on the readline() statement so actually this program will run forever unless manually stopped or
        it receives the "End" turn message from the game.
        """
        debug_write(BANNER_TEXT)

        while True:
            # Note: Python blocks and hangs on stdin. Can cause issues if connections aren't setup properly and may need to
            # manually kill this Python program.
            game_state_string = get_command()
            if "replaySave" in game_state_string:
                """
                This means this must be the config file. So, load in the config file as a json and add it to your AlgoStrategy class.
                """
                parsed_config = json.loads(game_state_string)
                self.on_game_start(parsed_config)
            elif "turnInfo" in game_state_string:
                state = json.loads(game_state_string)
                stateType = int(state.get("turnInfo")[0])
                if stateType == 0:
                    """
                    This is the game turn game state message. Algo must now print to stdout 2 lines, one for build phase one for
                    deploy phase. Printing is handled by the provided functions.
                    """
                    self.on_turn(game_state_string)
                elif stateType == 1:
                    """
                    If stateType == 1, this game_state_string string represents the results of an action phase
                    """
                    continue
                elif stateType == 2:
                    """
                    This is the end game message. This means the game is over so break and finish the program.
                    """
                    debug_write("Got end state quitting bot.")
                    break
                else:
                    """
                    Something is wrong? Recieved an incorrect or imporperly formatted string.
                    """
                    debug_write("Got unexpected string with turnInfo: {}".format(game_state_string))
            else:
                """
                Something is wrong? Recieved an incorrect or imporperly formatted string.
                """
                debug_write("Got unexpected string : {}".format(game_state_string))

File: community/c++-algo/include/GameLib/src/algoCore.cpp
/*
Description: Implementations for the algoCore header.
Last Modified: 06 Apr 2019
Author: Isaac Draper
*/

#include "algoCore.h"

namespace terminal {

    using json11::Json;

    /// Basic constructor for algoCore.
    AlgoCore::AlgoCore() {
        std::string error;
        config = Json::parse("{}", error);
        endOfGame = false;
    }

    /// A function that runs at the start of every game.
    /// Override this to perform initial setup at the start of the game, based on the configuration.
    /// @param configuration A Json object containing information about the game.
    void AlgoCore::onGameStart(Json configuration) {
        config = configuration;
    }

    /// Called when the engine expects input from the algo.
    /// It is called every turn and is passed a Json object containing
    /// the current game state, which can be used to initialize a new gameState.
    /// @param gameState A Json object containing the current game state.
    void AlgoCore::onTurn(Json gameState) {
        AlgoCore::submitDefaultTurn();
    }

    /// Submits a turn that does nothing.
    void AlgoCore::submitDefaultTurn() const {
        Util::sendCommand("[]");
        Util::sendCommand("[]");
    }

    /// Starts the main loop of the program.
    /// This will continue until it gets the end of state signal from the engine.
    /// If there is some error, it will never exit since it will continue
    /// to wait until it recieves input from the engine.
    /// If this happens it must be killed manually.
    void AlgoCore::start() {
        Util::debugWrite("Starting C++ Starter Algo");

        while (true) {
            try {
                Json gameState = Util::getCommand();

                if (gameState["turnInfo"] == nullptr) {
                    // Here we know it is the first turn, and the gameState object
                    // is information about the game, not an actual game state.
                    // Thus, we run the setup for the game.
                    
                    onGameStart(gameState);
                }
                else if (gameState["turnInfo"] != nullptr) {
                    int stateType = gameState["turnInfo"].array_items().at(0).int_value();

                    // We now make decisions based on what turn state it is.
                    switch (stateType) {
                        case 0: {
                            // This is when the engine expects to recieve what we are
                            // building for this turn.
                            // So we run the onTurn function.

                            onTurn(gameState);
                            break;
                        }
                        case 1: {
                            // This is a single frame from the engine.
                            // If you want to get information from the game every frame,
                            // you can create function to do that here.

                            break;
                        }
                        case 2: {
                            // This indicates that the game has ended.
                            // Since break simply takes us out of the switch,
                            // we use a variable to quite the while loop.

                            Util::debugWrite("GOT END OF STATE");
                            endOfGame = true;
                            break;
                        }
                        default: {
                            // This means we got an unexpected string.
                            // This technically should never happen since it would
                            // be caught by the Json parser.

                            Util::debugWrite("Unexpected state recieved: " + stateType);
                            break;
                        }
                    }

                    if (endOfGame) {
                        break;
                    }
                }
            }
            catch (UtilException e) {
                Util::debugWrite(e.what());
                continue;
            }
        }
    }

}
